WIP: support XPath

bicycle1885 opened the pull request:
This is a quick hack and still an experimental XPath support for XML documents.
If you are interested, I'll continue the job to finish this pull request.

``` julia
using LightXML

doc = parse_file("test/ex1.xml")
for n in evalxpath("//title", doc)
    @show n
end
```

output

```
n = <title lang="en">Everyday Italian</title>
n = <title lang="en">Harry Potter</title>
```

tkelman commented:
This looks like it would be useful. Right now xpath queries are supported in LibExpat.jl but not here, but I know the C API is there for them in libxml2.

bicycle1885 commented:
Thanks. I think supporting xpath in this module is useful because people would not want to switch libraries just to use xpath queries. libxml2 is often installed in many platforms and works out of box.
Now I added some tests and it works in some simple examples. Memory management and error handing are the remaining issues.

samoconnor commented:
xpath makes my brain hurt...

I've made an attempt at integrating my #39 with #38.

I got as far as this example in my test cases (edited here for brevity):

```julia
xml = """
<GetUserResponse xmlns="https://iam.amazonaws.com/doc/2010-05-08/">
  <GetUserResult>
    <User>
      <Arn>arn:aws:iam::012541411202:root</Arn>
      <UserId>012541411202</UserId>
    </User>
  </GetUserResult>
</GetUserResponse>
"""

@time @test parse_string(xml)["GetUserResult"]["User"]["Arn"] == "arn:aws:iam::012541411202:root"

@time @test strip(content(
            evalxpath("//GetUserResult/User/Arn", parse_string(xml))[1]
      )) == "arn:aws:iam::012541411202:root"
```

```julia
signal (11): Segmentation fault: 11
getindex at /Users/sam/.julia/v0.4/LightXML/src/xpath.jl:70
```

This works if I remove the `xmlns=` attribute.

This seems to be because of libxml's stubborn compliance to the letter of the xpath standard. 
See: http://plasmasturm.org/log/259/, quote:
> You cannot define a default namespace for XPath, period, don’t try you can’t, the XPath spec does not allow it.

Another example, adding a dummy `xmlns=` to your wikimedia example like this also causes a failure:

```julia
xdoc = parse_string("""
<?xml version="1.0" encoding="utf-8"?>
<wikimedia xmlns="http://foobar.com">
  <projects>
...
```

bicycle1885 commented:
@samoconnor Your brackets-based indexing may be handy, but I believe that it cannot replace the xpath entirely. Is there any equivalent of `//edition` query? 

I noticed namespace things could cause problems, they are in my todo list.

samoconnor commented:
@bicycle1885 No doubt xpath is a complex and powerful beast and can do way more than my simple hack for traversing elements with `[]` syntax.

It would be trivial to add a `["*"]` syntax to my `getindex()` code, but my preference is to keep it really simple. It's mostly intended for parsing really simple XML documents (like web service responses).

In fact, the first hack I came up with was equivalent to `//`, but in the end I found that it is more robust to match the whole path. Particularly with web services where the other end could change their format from time to time it is important to not accidentally match something unexpected.

Note also, I've posted some benchmarks over at #39.

nalimilan commented:
@samoconnor I looks unlikely to me that the "stubborn compliance" of libxml2 would cause that call to crash. At worst, it should return `NULL`. I guess when `xmlns=` is added, the library changes its behavior and your `ccall` is not longer valid. For example, libxml2 might try to write the namespace information to one of the fields of the `xmlXPathContext`, and for some reason this wouldn't work.

nalimilan commented on src/xpath.jl:
```diff
@@ -0,0 +1,92 @@
+    function XPathObject(ptr::Ptr)
+        return new(ptr)
+    end
+end
```
You need to add a finalizer calling `xmlXPathFreeObject`.

nalimilan commented on src/xpath.jl:
```diff
@@ -0,0 +1,92 @@
+        end
+        return new(ptr)
+    end
+end
```
You need to add a finalizer calling `xmlXPathFreeContext`.

nalimilan commented:
I just saw the crash happens in `getindex`. It's just that the`xmlXPathEvalExpression` call returns an empty set, so you cannot access the first element. You should simply improve the checks.

bicycle1885 commented:
@nalimilan Thank you for your feedbacks. I added finalizers to free resources and a bounds checker. No longer segfaults happen.
And I added a `registerns!` function to register a namespace to the xpath context, which enables to query elements with namespaces.

@samoconnor 
> No doubt xpath is a complex and powerful beast and can do way more than my simple hack for traversing elements with [] syntax.

XPath has well-defined specs and a standard way to query elements of a complicated document. Using XPath, we can write fairly complicated queries in a single line and extract deeply nested elements. Your use case and my use case would be totally different. I'd like to parse and query elements in XML documents of >100MiB size with a mixture of several different namespaces. See the go.owl file in this link http://geneontology.org/page/download-ontology.

bicycle1885 commented:
I'm going to implement this in https://github.com/bicycle1885/EzXML.jl.
